exchange-v2/test/v2/EIP712MetaTransaction.test.js

```diff
@@ -0,0 +1,689 @@
+
+  return {r, s, v, functionSignature};
+}
+
+contract("EIP712MetaTransaction", function ([_, owner, account1]) {
+  let testContract
+  let testContractSimple;
+  let transferProxy;
+  let erc20TransferProxy;
+  let t1;
+  let t2;
+  let community = account1;
+  let left;
+  let right;
+
+  before('before', async function () {
+    transferProxy = await TransferProxyTest.new();
+    erc20TransferProxy = await ERC20TransferProxyTest.new();
+    royaltiesRegistry = await TestRoyaltiesRegistry.new();
+    testContract = await deployProxy(ExchangeV2, [transferProxy.address, erc20TransferProxy.address, 300, community, royaltiesRegistry.address], { initializer: "__ExchangeV2_init" });
+    testingSimpleContract = await deployProxy(ExchangeSimpleV2, [transferProxy.address, erc20TransferProxy.address], { initializer: "__ExchangeSimpleV2_init" });
+    t1 = await TestERC20.new();
+    t2 = await TestERC20.new();
+
+    domainData = {
+      name: "ExchangeV2",
+      version: "1",
+      verifyingContract: testContract.address,
+      chainId: 1337
+    };
+
+    await t1.mint(owner, 100);
+    await t1.approve(erc20TransferProxy.address, 10000000, { from: owner });
+    left = Order(publicKey, Asset(ERC20, enc(t1.address), 100), ZERO, Asset(ERC20, enc(t2.address), 200), 1, 0, 0, "0xffffffff", "0x");
+    right = Order(owner, Asset(ERC20, enc(t1.address), 100), ZERO, Asset(ERC20, enc(t2.address), 200), 1, 0, 0, "0xffffffff", "0x");
+  });
+
+  describe("Check Methods", function () {
+  	it("upgrade, which use MetaTransaction  works", async () => {
+  		const wrapper = await ExchangeSimpleV2_MetaTx.at(testingSimpleContract.address);
+  		await expectThrow(
+  			wrapper.getNonce(ZERO_ADDRESS)
+  		);
+
+  		await upgradeProxy(testingSimpleContract.address, ExchangeSimpleV2_MetaTx);
+  		assert.equal(await wrapper.getNonce(ZERO_ADDRESS), 0);
+  	});
+
+    it("Should be able to send transaction successfully, and check Event, that emit from method, execute as MetaTx", async () => {
+      let nonce = await testContract.getNonce(publicKey);
+
+      let {
+        r,
+        s,
+        v,
+        functionSignature
+      } = await getTransactionData(nonce, cancelAbi, [left]);
+
+      let sendTransactionData = web3Abi.encodeFunctionCall(
+        executeMetaTransactionABI,
+        [publicKey, functionSignature, r, s, v]
+      );
+//        Way №1 call transaction
+//        await testContract.sendTransaction({
+//            value: 0,
+//            from: owner,
+//            gas: 500000,
+//            data: sendTransactionData
+//        });
+//        Way №2 call transaction
+      let resultExecMataTx  = await testContract.executeMetaTransaction(publicKey, functionSignature, r, s, v, {from: owner});
+      let orderMakerAddress;
+      truffleAssert.eventEmitted(resultExecMataTx, 'Cancel', (ev) => {
+       	orderMakerAddress = ev.maker;
+        return true;
+      });
+      console.log("orderMakerAddress:"+orderMakerAddress);
```

Comment:
not sure if we need console.log in tests

Reply:
Done